orttraining/orttraining/python/training/checkpoint.py

```diff
@@ -342,13 +371,76 @@ def aggregate_checkpoints(paths, pytorch_format=True):
                 state_dict[_utils.state_dict_user_dict_key()] = rank_state_dict[_utils.state_dict_user_dict_key()]
 
-    # reshape all the sharded tensors based on the original dimensions stored in sharded_states_original_dims
-    _reshape_states(sharded_states_original_dims, state_dict, loaded_mixed_precision)
+    if reshape_states:
+        # reshape all the sharded tensors based on the original dimensions stored in sharded_states_original_dims
+        _reshape_states(sharded_states_original_dims, state_dict, loaded_mixed_precision)
 
     # return a flat structure for PyTorch model in case pytorch_format is True
     # else return the hierarchical structure for ORTTrainer
     return _to_pytorch_format(state_dict) if pytorch_format else state_dict
 
+def aggregate_checkpoints(paths, pytorch_format=True):
+    """Aggregate checkpoint files and return a single state dictionary
+
+    Aggregates checkpoint files specified by paths and laods the checkpoint file one at a time merging
+    them into a single state dictionary.
+    The checkpoint files represented by paths must be saved through ORTTrainer.save_checkpoint() function.
+    The schema of the state_dict returned will be in the same as the one returned by ORTTrainer.state_dict()
+
+    Args:
+        paths: list of more than one file represented as strings where the checkpoint is saved
+        pytorch_format: boolean flag to select either ONNX Runtime or PyTorch state schema of the returned state_dict
+    Returns:
+        state_dict that can be loaded into an ORTTrainer or into a PyTorch model
+    """
+
+    # order the paths in ascending order of ranks
+    ordered_paths = _order_paths(paths)
+    loaded_trainer_options = _checkpoint_storage.load(paths[0], key=_utils.state_dict_trainer_options_key())
+    D_size = _utils.state_dict_trainer_options_data_parallel_size_key()
+    H_size = _utils.state_dict_trainer_options_horizontal_parallel_size_key()
+    world_size = _utils.state_dict_trainer_options_world_size_key()
+
+    D_size = loaded_trainer_options[D_size]
+    H_size = loaded_trainer_options[H_size]
+    world_size = loaded_trainer_options[world_size]
+    D_groups, H_groups = _get_parallellism_groups(D_size, H_size, world_size)
+
+    combine_zero = loaded_trainer_options[_utils.state_dict_trainer_options_zero_stage_key()] > 0
+    combine_megatron = len(H_groups[0]) > 1
+
+    aggregate_state = None
+    if combine_zero and combine_megatron:
+        sharded_states_original_dims = {}
+        save_dir = tempfile.mkdtemp()
+        os.makedirs(save_dir, exist_ok = True) 
+
+        aggregate_data_checkpoint_files = []
+        # combine for Zero over data groups and save to temp file
+        for group_id in range(len(D_groups)):
+            aggregate_state_dict = _aggregate_checkpoints([ordered_paths[i] for i in D_groups[group_id]], sharded_states_original_dims, reshape_states=False, add_partition_info = True, pytorch_format=False)
```

Comment:
after this aggregation, the state dict has ```trainer_options``` subdict which have the ```worls_size```, ```D_size``` and ```H_size``` set to 1. Is that expected? Since we are only aggregating one of two (ZeRO or Megatron) away, should the ```trainer_options``` need to be preserved for the remaining aggregation?

Reply:
the trainer options  D_size and H_size are needed only in the beginning in aggregate_states in order to get the necessary order of aggregation, and are not used further, so it is okay to set them to 1 which is the expected aggregated value